ch4: Using GenQ in pack buffer management for CH4 AM small messages

yfguo opened the pull request:
## Pull Request Description

This PR uses GenQ private to implement the pack buffer management for CH4 AM small messages.

The first 7 commits are prerequisites. The first three prereq commits make the OFI AM receive buffer, RDMA buffer and UCX AM recv buffer (for all msg sizes) to use GPU registered memory so that the data can be efficiently copy/unpack to user buffer on GPU. The next two prereq commits fixes the eager limit query of SHM and NETMOD. The last two prereq commits add the interface for querying the size of eager buffer in SHM and the size of pack buffer in NETMOD eager send. These information are used to determine if the receive side pack buffer is large enough at init time.

This last three commits involve three pieces of codes:

i. MPIDIG AM recv for small unexpected messages
ii. OFI AM send for small messages

The existing code use MPL_gpu_malloc_host to allocate the pack buffer for (i) and (ii). This brings the memory registration overhead to the small message path.

To address these issues, this PR did three changes:

**1. Using GenQ for pack buffer in AM recv of small unexpected messages.**
In this case, the pack buffer size (cell size) is 17 * 4KB to match the greater between the POSIX iQueue cell size and the OFI eager threshold. This is because the AM recv code is used by both SHM and NETMOD, therefore pack buffer size need to be large enough for the small message in both. The cell size is check against the eager buffer size of SHM and the pack buffer size of NETMOD eager send at init time to make sure they are large enough.

The max number of cells in this unexpected pack buffer pool is INT_MAX.

In the current implementation, if the incoming data is larger than the cell size, we consider this is the large message case and still do a MPL_gpu_malloc_host to get the pack buffer. Currently, this is path is used when a large message is sending from UCX AM which does not differentiate between small and large messages. 

**2. Using GenQ for pack buffer in AM send of small messages in OFI.**
In this case, the pack buffer size (cell size) is 16KB (eager threshold of OFI). The pool is created with limited number of buffers. This limit is set in CVAR. If pool has no free buffer, the posting of am_isend will be deferred and retried from progress engine.

If there is a deferred am isend waiting, all subsequent am isend will be deferred as well to maintain the ordering.

For large message, we still need to allocate the pack buffer for the data with MPL_gpu_malloc_host. This receive side will do RDMA read from this buffer.

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/ofi

arakashic commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

pavanbalaji commented:
Testing UCX as well, since there were changes to the upper layers.

test:mpich/ch4/ucx

yfguo commented:
Tests are still failing. I will work on it later tonight.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
This should fixed the buffer overwriting. Fixing the segfault and rma issues now.

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
Turns out the LMT protocol has an uninitialized field bug which cause the collectives to fail.

test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
1. Collective failures in default, debug, strict, no-inline, external builds. This is due to an uninitialized match_req field in the current LMT protocol implementation. Fixed

2. RMA failures in direct-nm build. This is due to buggy implementation of deferred isend in OFI.

3. Multithread failures in am-only build. This is actually due to the received side AM running out of buffer from GenQ. This is exposed in the multithreaded tests because the large amount of unexpected message using up all buffer. The solution is adding deferred target_msg callback handling (only one callback has this problem---MPIDIG_send_target_msg_cb). 

4. Datatype and RMA (putfidx) failures in am-only build. Still locating the problem. Fixed now.

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ofi

yfguo commented:
test:mpich/ch4/ucx

yfguo commented:
test:mpich/ch4/most

yfguo commented on src/mpid/ch4/netmod/ofi/ofi_am_impl.h:
```diff
@@ -361,11 +396,18 @@ static inline int MPIDI_OFI_do_am_isend(int rank,
+            }
+        } else {
+            /* only for large message RDMA read */
+            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
```
This is needed only for the large message RDMA read at the moment as we do not have a pipelined AM protocol for large messages.

pavanbalaji replied:
We do have it for shared memory (the LMT protocol), just not for network communication.  The RDMA read based protocol makes sense when the data is being sent directly from the user buffer.  But when there's a temporary buffer involved (e.g., for noncontiguous data, or for contiguous data on GPUs when GPU Direct is not available), we should switch to the LMT protocol, as that's more appropriate.

yfguo replied:
I am working on bring the SHM LMT to netmod.

yfguo commented:
test:mpich/ch4/most

pavanbalaji commented on src/mpid/ch4/generic/am/mpidig_am_init.c:
```diff
@@ -8,21 +8,78 @@
 
+static void *host_alloc_buffer_registered(uintptr_t size)
+{
+    void *ptr = MPL_malloc(size, MPL_MEM_BUFFER);
```
```suggestion
    void *ptr = MPL_malloc(size, MPL_MEM_BUFFER);
    MPIR_Assert(ptr);
```

yfguo replied:
I think this wouldn't be necessary, as the GenQ already check for NULL pointer and throws an error.

pavanbalaji replied:
But you call buffer registration immediately after that, which won't work on NULL pointers.

yfguo replied:
fixed

pavanbalaji commented on src/mpid/ch4/generic/am/mpidig_am_init.c:
```diff
@@ -70,6 +127,17 @@ int MPIDIG_am_init(void)
+                                                      &MPIDI_global.unexp_pack_buf_pool);
+    MPIR_ERR_CHECK(mpi_errno);
+    MPIDU_genq_private_pool_set_warning_limit(MPIDI_global.unexp_pack_buf_pool,
+                                              MPIR_CVAR_CH4_MAX_AM_UNEXPECTED_PACK_BUFFERS_SIZE_BYTE);
```
I don't particularly like this function.  It's very specific to this use case.  I'd recommend simply dropping this functionality.

yfguo replied:
I see, then for the size warning, I can just tracking it inside CH4.

pavanbalaji replied:
I don't think there's an easy way to track it.  That's why I recommended that you simply drop that functionality.

yfguo replied:
dropped

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

pavanbalaji commented on src/mpid/ch4/generic/am/mpidig_am_recv.h:
```diff
@@ -9,22 +9,70 @@
+                                                           MPIDIG_REQUEST(rreq, buffer));
+            if (MPIDIG_REQUEST(rreq, buffer) == NULL) {
+                mpi_errno = MPI_ERR_OTHER;
+            }
```
Why is this error code being overridden compared to what `MPIDU_genq_private_pool_alloc_cell` already set?  Shouldn't you just jump to fn_fail if this happens, instead?

pavanbalaji commented on src/mpid/ch4/generic/am/mpidig_am_recv.c:
```diff
@@ -0,0 +1,47 @@
+    mpi_errno = MPIDIG_prepare_recv_req(dreq->rank, dreq->tag, dreq->context_id, dreq->buf,
+                                        dreq->count, dreq->datatype, dreq->rreq);
+    if (mpi_errno != MPI_SUCCESS) {
+        *status = MPIDIG_AM_IRECV_DEFERRED;
```
Can you use a different variable name than "status", to avoid confusion with `MPI_Status`?

pavanbalaji commented on src/mpid/ch4/generic/am/mpidig_am_recv.c:
```diff
@@ -0,0 +1,47 @@
+
+    MPIR_Comm_add_ref(root_comm);
+    MPIDIG_enqueue_posted(dreq->rreq, &MPIDIG_COMM(root_comm, posted_list));
+    *status = MPI_SUCCESS;
```
Why is status using `MPI_SUCCESS`?  It's should use an enum that's consistent with `MPIDIG_AM_IRECV_DEFERRED`.  Also, it uses a hack that assumes that `MPIDIG_AM_IRECV_DEFERRED` is always non-zero, so it doesn't conflict with `MPI_SUCCESS`.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_progress.c:
```diff
@@ -7,6 +7,34 @@
+{
+
+    int mpi_errno = MPI_SUCCESS;
+    int status = MPI_SUCCESS;
```
Can you not use `status` in all these commits?

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_am_impl.c:
```diff
@@ -0,0 +1,66 @@
+                                    send_buf, dreq->data_sz, dreq->sreq);
+    }
+    MPIR_ERR_CHECK(mpi_errno);
+    *status = MPI_SUCCESS;
```
`status` should not be set to an MPI error code in one place and to an enum in another place.

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most

yfguo commented:
test:mpich/ch4/most
